File: output/amazon_q/python/hard/q44-count-subtrees-with-max-distance-between-cities.py
# https://leetcode.com/problems/count-subtrees-with-max-distance-between-cities/
# Problem: Count Subtrees With Max Distance Between Cities

#Problem Description:
#There are n cities numbered from 1 to n. You are given an array edges of size n-1, where edges[i] = [ui, vi] represents a bidirectional edge between cities ui and vi. There exists a unique path between each pair of cities. In other words, the cities form a tree.
#A subtree is a subset of cities where every city is reachable from every other city in the subset, where the path between each pair passes through only the cities from the subset. Two subtrees are different if there is a city in one subtree that is not present in the other.
#For each d from 1 to n-1, find the number of subtrees in which the maximum distance between any two cities in the subtree is equal to d.
#Return an array of size n-1 where the dth element (1-indexed) is the number of subtrees in which the maximum distance between any two cities is equal to d.
#Notice that the distance between the two cities is the number of edges in the path between them.

# Constraints:
# 2 <= n <= 15
# edges.length == n-1
# edges[i].length == 2
# 1 <= ui, vi <= n
# All pairs (ui, vi) are distinct.

#Code Structure:

# First generated code

import collections
class Solution(object):
    def countSubgraphsForEachDiameter(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[int]
        """
        graph = collections.defaultdict(list)
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)

        def dfs(mask, u, parent):
            mask ^= 1 << u
            max_depth = 0
            for v in graph[u]:
                if mask & (1 << v) and v != parent:
                    depth = dfs(mask, v, u)
                    if max_depth == 0:
                        max_depth = depth
                    elif depth != 0:
                        max_depth = max(max_depth, depth)
            return max_depth + 1

        def get_diameter(mask):
            if mask == 0:
                return 0
            max_depth = 0
            for u in range(n):
                if mask & (1 << u):
                    depth = dfs(mask, u, -1)
                    if max_depth == 0:
                        max_depth = depth
                    elif depth != 0:
                        max_depth = max(max_depth, depth)
                    break
            return max_depth - 1

        result = [0] * (n - 1)
        for mask in range(1, 1 << n):
            diameter = get_diameter(mask)
            if diameter > 0:
                result[diameter - 1] += 1
        return result                                                      
    

# Debugged Code

from collections import defaultdict, deque

class Solution(object):
    def countSubgraphsForEachDiameter(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[int]
        """
        graph = defaultdict(list)
        for u, v in edges:
            graph[u-1].append(v-1)
            graph[v-1].append(u-1)
        
        result = [0] * (n - 1)
        
        for mask in range(1, 1 << n):
            if bin(mask).count('1') < 2:
                continue
            
            cities = [i for i in range(n) if mask & (1 << i)]
            
            if not self.isConnected(cities, graph, mask):
                continue
            
            max_dist = self.getMaxDistance(cities, graph)
            if max_dist > 0:
                result[max_dist - 1] += 1
        
        return result
    
    def isConnected(self, cities, graph, mask):
        if not cities:
            return False
        
        visited = set()
        queue = deque([cities[0]])
        visited.add(cities[0])
        
        while queue:
            node = queue.popleft()
            for neighbor in graph[node]:
                if (mask & (1 << neighbor)) and neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
        
        return len(visited) == len(cities)
    
    def getMaxDistance(self, cities, graph):
        max_dist = 0
        
        for i in range(len(cities)):
            for j in range(i + 1, len(cities)):
                dist = self.bfs(cities[i], cities[j], graph, set(cities))
                max_dist = max(max_dist, dist)
        
        return max_dist
    
    def bfs(self, start, end, graph, city_set):
        if start == end:
            return 0
        
        visited = set([start])
        queue = deque([start])
        dist = 0
        
        while queue:
            size = len(queue)
            dist += 1
            
            for _ in range(size):
                curr = queue.popleft()
                
                for neighbor in graph[curr]:
                    if neighbor == end:
                        return dist
                    if neighbor in city_set and neighbor not in visited:
                        visited.add(neighbor)
                        queue.append(neighbor)
        
        return -1                                                      
    

File: output/amazon_q/java/hard/q44-count-subtrees-with-max-distance-between-cities.java
/** https://leetcode.com/problems/count-subtrees-with-max-distance-between-cities */
// Problem: Count Subtrees With Max Distance Between Cities

//Problem Description:
//There are n cities numbered from 1 to n. You are given an array edges of size n-1, where edges[i] = [ui, vi] represents a bidirectional edge between cities ui and vi. There exists a unique path between each pair of cities. In other words, the cities form a tree.
//A subtree is a subset of cities where every city is reachable from every other city in the subset, where the path between each pair passes through only the cities from the subset. Two subtrees are different if there is a city in one subtree that is not present in the other.
//For each d from 1 to n-1, find the number of subtrees in which the maximum distance between any two cities in the subtree is equal to d.
//Return an array of size n-1 where the dth element (1-indexed) is the number of subtrees in which the maximum distance between any two cities is equal to d.
//Notice that the distance between the two cities is the number of edges in the path between them.

// Constraints:
// 2 <= n <= 15
// edges.length == n-1
// edges[i].length == 2
// 1 <= ui, vi <= n
// All pairs (ui, vi) are distinct.

//Code Structure:

// First generated code

class Solution {
    public int[] countSubgraphsForEachDiameter(int n, int[][] edges) {
        int[] res = new int[n - 1];
        int[][] graph = new int[n][n];
        for (int[] edge : edges) {
            graph[edge[0] - 1][edge[1] - 1] = 1;
            graph[edge[1] - 1][edge[0] - 1] = 1;
        }
        for (int i = 0; i < n; i++) {
            graph[i][i] = 1;
        }
        for (int i = 1; i <= n - 1; i++) {
            boolean[] visited = new boolean[n];
            dfs(graph, visited, 0, i, res);
        }
        return res;
        
    }
}

// Debugged Code
    
class Solution {
    public int[] countSubgraphsForEachDiameter(int n, int[][] edges) {
        int[] res = new int[n - 1];
        int[][] graph = new int[n][n];
        for (int[] edge : edges) {
            graph[edge[0] - 1][edge[1] - 1] = 1;
            graph[edge[1] - 1][edge[0] - 1] = 1;
        }
        
        for (int mask = 1; mask < (1 << n); mask++) {
            if (Integer.bitCount(mask) < 2) continue;
            int maxDist = getMaxDistance(graph, mask, n);
            if (maxDist > 0) res[maxDist - 1]++;
        }
        return res;
    }
    
    private int getMaxDistance(int[][] graph, int mask, int n) {
        int[] cities = new int[Integer.bitCount(mask)];
        int idx = 0;
        for (int i = 0; i < n; i++) {
            if ((mask & (1 << i)) != 0) cities[idx++] = i;
        }
        
        if (!isConnected(graph, cities)) return -1;
        
        int maxDist = 0;
        for (int i = 0; i < cities.length; i++) {
            for (int j = i + 1; j < cities.length; j++) {
                maxDist = Math.max(maxDist, bfs(graph, cities[i], cities[j], mask));
            }
        }
        return maxDist;
    }
    
    private boolean isConnected(int[][] graph, int[] cities) {
        boolean[] visited = new boolean[cities.length];
        dfs(graph, cities, visited, 0);
        for (boolean v : visited) {
            if (!v) return false;
        }
        return true;
    }
    
    private void dfs(int[][] graph, int[] cities, boolean[] visited, int idx) {
        visited[idx] = true;
        for (int i = 0; i < cities.length; i++) {
            if (!visited[i] && graph[cities[idx]][cities[i]] == 1) {
                dfs(graph, cities, visited, i);
            }
        }
    }
    
    private int bfs(int[][] graph, int start, int end, int mask) {
        if (start == end) return 0;
        boolean[] visited = new boolean[graph.length];
        int[] queue = new int[graph.length];
        int front = 0, rear = 0;
        queue[rear++] = start;
        visited[start] = true;
        int dist = 0;
        
        while (front < rear) {
            int size = rear - front;
            dist++;
            for (int i = 0; i < size; i++) {
                int curr = queue[front++];
                for (int next = 0; next < graph.length; next++) {
                    if (graph[curr][next] == 1 && !visited[next] && (mask & (1 << next)) != 0) {
                        if (next == end) return dist;
                        visited[next] = true;
                        queue[rear++] = next;
                    }
                }
            }
        }
        return -1;
    }
}
    